File: backend/src/main/java/com/financialplanner/service/TransactionService.java
package com.financialplanner.service;

import com.financialplanner.model.Transaction;
import com.financialplanner.repository.TransactionRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

/**
 * Service layer for Transaction CRUD operations with business logic and validation
 */
@Service
public class TransactionService {

    @Autowired
    private TransactionRepository transactionRepository;

    /**
     * Create a new transaction with validation
     */
    public Transaction createTransaction(Transaction transaction) {
        validateTransaction(transaction);
        return transactionRepository.save(transaction);
    }

    /**
     * Retrieve a transaction by ID
     */
    public Optional<Transaction> getTransaction(String id) {
        return transactionRepository.findById(id);
    }

    /**
     * Retrieve all transactions
     */
    public List<Transaction> getAllTransactions() {
        return transactionRepository.findAll();
    }

    /**
     * Retrieve transactions by category
     */
    public List<Transaction> getTransactionsByCategory(String category) {
        return transactionRepository.findByCategory(category);
    }

    /**
     * Retrieve transactions within a date range
     */
    public List<Transaction> getTransactionsByDateRange(LocalDate startDate, LocalDate endDate) {
        return transactionRepository.findByDateBetween(startDate, endDate);
    }

    /**
     * Retrieve transactions by type (INCOME or EXPENSE)
     */
    public List<Transaction> getTransactionsByType(Transaction.TransactionType type) {
        return transactionRepository.findByType(type);
    }

    /**
     * Update an existing transaction
     */
    public Optional<Transaction> updateTransaction(String id, Transaction transaction) {
        validateTransaction(transaction);
        return transactionRepository.update(id, transaction);
    }

    /**
     * Delete a transaction
     */
    public boolean deleteTransaction(String id) {
        return transactionRepository.delete(id);
    }

    /**
     * Get summary of transactions for a specific month and year
     */
    public TransactionSummary getMonthSummary(int year, int month) {
        LocalDate startDate = LocalDate.of(year, month, 1);
        LocalDate endDate = startDate.plusMonths(1).minusDays(1);

        List<Transaction> transactions = getTransactionsByDateRange(startDate, endDate);

        double totalIncome = transactions.stream()
                .filter(t -> t.getType() == Transaction.TransactionType.INCOME)
                .mapToDouble(Transaction::getAmount)
                .sum();

        double totalExpense = transactions.stream()
                .filter(t -> t.getType() == Transaction.TransactionType.EXPENSE)
                .mapToDouble(Transaction::getAmount)
                .sum();

        return TransactionSummary.builder()
                .year(year)
                .month(month)
                .totalIncome(totalIncome)
                .totalExpense(totalExpense)
                .netAmount(totalIncome - totalExpense)
                .transactionCount(transactions.size())
                .build();
    }

    /**
     * Validate transaction data
     */
    private void validateTransaction(Transaction transaction) {
        if (transaction.getDate() == null) {
            throw new IllegalArgumentException("Transaction date cannot be null");
        }
        if (transaction.getAmount() == null || transaction.getAmount() <= 0) {
            throw new IllegalArgumentException("Transaction amount must be greater than 0");
        }
        if (transaction.getCategory() == null || transaction.getCategory().trim().isEmpty()) {
            throw new IllegalArgumentException("Transaction category cannot be empty");
        }
        if (transaction.getType() == null) {
            throw new IllegalArgumentException("Transaction type must be INCOME or EXPENSE");
        }
        if (transaction.getDescription() == null || transaction.getDescription().trim().isEmpty()) {
            throw new IllegalArgumentException("Transaction description cannot be empty");
        }
    }
}

File: frontend/src/app/services/transaction.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Transaction, TransactionSummary, AnalysisResponse } from '../models/transaction.model';

@Injectable({
  providedIn: 'root'
})
export class TransactionService {
  private apiUrl = 'http://localhost:8080/api/transactions';

  constructor(private http: HttpClient) {}

  /**
   * Get all transactions
   */
  getAllTransactions(): Observable<Transaction[]> {
    return this.http.get<Transaction[]>(this.apiUrl);
  }

  /**
   * Get a single transaction by ID
   */
  getTransaction(id: string): Observable<Transaction> {
    return this.http.get<Transaction>(`${this.apiUrl}/${id}`);
  }

  /**
   * Create a new transaction
   */
  createTransaction(transaction: Transaction): Observable<Transaction> {
    return this.http.post<Transaction>(this.apiUrl, transaction);
  }

  /**
   * Update an existing transaction
   */
  updateTransaction(id: string, transaction: Transaction): Observable<Transaction> {
    return this.http.put<Transaction>(`${this.apiUrl}/${id}`, transaction);
  }

  /**
   * Delete a transaction
   */
  deleteTransaction(id: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  /**
   * Get transactions filtered by category
   */
  getTransactionsByCategory(category: string): Observable<Transaction[]> {
    let params = new HttpParams().set('category', category);
    return this.http.get<Transaction[]>(`${this.apiUrl}/filter/category`, { params });
  }

  /**
   * Get transactions filtered by date range
   */
  getTransactionsByDateRange(startDate: string, endDate: string): Observable<Transaction[]> {
    let params = new HttpParams()
      .set('startDate', startDate)
      .set('endDate', endDate);
    return this.http.get<Transaction[]>(`${this.apiUrl}/filter/date`, { params });
  }

  /**
   * Get transactions filtered by type
   */
  getTransactionsByType(type: 'INCOME' | 'EXPENSE'): Observable<Transaction[]> {
    let params = new HttpParams().set('type', type);
    return this.http.get<Transaction[]>(`${this.apiUrl}/filter/type`, { params });
  }

  /**
   * Get monthly summary
   */
  getMonthSummary(year: number, month: number): Observable<TransactionSummary> {
    let params = new HttpParams()
      .set('year', year.toString())
      .set('month', month.toString());
    return this.http.get<TransactionSummary>(`${this.apiUrl}/summary`, { params });
  }

  /**
   * Request AI-powered analysis for a month/year
   */
  getAnalysis(year: number, month: number): Observable<AnalysisResponse> {
    let params = new HttpParams()
      .set('year', year.toString())
      .set('month', month.toString());
    // note: analysis endpoint is served at root /analysis rather than under /transactions
    return this.http.get<AnalysisResponse>('http://localhost:8080/analysis', { params });
  }
}
